bundle/edu.gemini.itc/src/main/scala/edu/gemini/itc/service/ItcServiceImpl.scala

```diff
@@ -1,48 +1,64 @@
+ *
+ * Note that all results are repacked in simplified Scala case classes in order not to leak out any of the
+ * implementation details of the underlying ITC functionality.
  */
 class ItcServiceImpl extends ItcService {
 
   import ItcService._
 
   def calculate(source: SourceDefinition, obs: ObservationDetails, cond: ObservingConditions, tele: TelescopeDetails, ins: InstrumentDetails): Result = try {
-    ins match {
-      case i: GmosParameters  => calculateGmos(source, obs, cond, tele, i)
-      case _                  => throw new NotImplementedError // TODO: no other instruments are implemented yet
-    }
+
+    if (obs.getMethod.isImaging)  calculateImaging(source, obs, cond, tele, ins)
+    else                          calculateSpectroscopy(source, obs, cond, tele, ins)
+
   } catch {
     // TODO: for now in most cases where a validation problem should be reported to the user the ITC code throws an exception instead
     case e: Throwable =>
+      e.printStackTrace()
       ItcResult.forException(e)
   }
 
-  private def calculateGmos(source: SourceDefinition, obs: ObservationDetails, cond: ObservingConditions, tele: TelescopeDetails, ins: GmosParameters): Result = {
-    val recipe = new GmosRecipe(source, obs, cond, ins, tele)
-    // TODO: we can simplify this once all recipes have a calculate method (instead of writeOutput())
-    val results: Array[ItcCalcResult] = obs.getMethod match {
-      case m: Imaging =>
-          // Repack the result in an immutable and simplified Scala case class
-          // (We don't want to leak out any of the internal ITC craziness here and it is also a good way
-          // to keep the service independent from the actual implementation.)
-          recipe.calculateImaging().map {
-            case r: ImagingResult => r.is2nCalc match {
-              case i: ImagingS2NMethodACalculation  => ItcImagingResult(i.singleSNRatio(), i.totalSNRatio(), r.peakPixelCount)
-              case _                                => throw new NotImplementedError()
-            }
-          }
-
-      case s: Spectroscopy =>
-        throw new NotImplementedError()
+  // === Imaging
+
+  private def calculateImaging(source: SourceDefinition, obs: ObservationDetails, cond: ObservingConditions, tele: TelescopeDetails, ins: InstrumentDetails): Result =
+    ins match {
+      case i: AcquisitionCamParameters  => imagingResult(new AcqCamRecipe(source, obs, cond, tele, i))
+      case i: Flamingos2Parameters      => imagingResult(new Flamingos2Recipe(source, obs, cond, i, tele))
+      case i: GmosParameters            => imagingArrayResult(new GmosRecipe(source, obs, cond, i, tele))
+      case i: GsaoiParameters           => imagingResult(new GsaoiRecipe(source, obs, cond, i, tele))
+      case i: NiriParameters            => imagingResult(new NiriRecipe(source, obs, cond, i, tele))
+      case _                            => throw new NotImplementedError
```

Comment:
I changed some of the error handling here to makes use of ```Validation``` however I kept the lower level error message as an exception. The point here is that the error/validation/problem handling in the original Java code uses exceptions everywhere which are caught on the highest level (i.e. servlet and now the service implementation). Changing this might be worthwhile for some parts but probably not for everything (just because it's probably not worth the time to rewrite all of that in Scala). I haven't put any effort into this yet because it doesn't seem to be a hight priority, however I might change some of the input parameter checks at some point in the future.